modules/solid_mechanics/src/bcs/DirectDirichletBCBase.C

```diff
@@ -0,0 +1,53 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "DirectDirichletBCBase.h"
+#include "NonlinearSystemBase.h"
+
+InputParameters
+DirectDirichletBCBase::validParams()
+{
+  InputParameters params = NodalBC::validParams();
+  params.addParam<TagName>("mass_matrix_tag", "mass", "The tag for the mass matrix");
+
+  return params;
+}
+
+DirectDirichletBCBase::DirectDirichletBCBase(const InputParameters & parameters)
+  : NodalBC(parameters), _mass_matrix(nullptr)
+{
+}
+
+void
+DirectDirichletBCBase::initialSetup()
+{
+  _mass_matrix = &_sys.system().get_matrix("System Matrix");
+}
+
+Real
+DirectDirichletBCBase::computeQpResidual()
+{
+  // Get dof for current var
+  const auto dofnum =
+      _current_node->dof_number(_sys.number(), variable().number(), _sys.dofMap().sys_number());
+
+  // Get lumped component of mass matrix
+  Real masslump = 0;
+  for (const auto j : make_range(_mass_matrix->n()))
+    masslump += (*_mass_matrix)(dofnum, j);
+
+  // Compute residual to enforce BC
+  // This is the force required to enforce the BC in a central difference scheme
+  Real avg_dt = (_dt + _dt_old) / 2;
+  Real resid =
+      (computeQpValue() - _u[_qp]) / (avg_dt * _dt) - (*_sys.solutionUDotOld())(dofnum) / avg_dt;
```

Comment:
I'd like to understand the derivation, i.e. why the denominator is (avg_dt * _dt)

Reply:
CD Solution update:
`a_n = M^-1 F` 
`v_(n+1/2) = v_(n-1/2) + avg_dt * a_n`
`d_(n+1) = d_n + dt * v_(n+1/2)`
Here u_(n+1) can be thought of as the boundary value that needs to be enforced, So u_(n+1) = u_g
Then combine the solution update into one equation,

`u_g = u_n + _dt * (v_(n-1/2) + avg_dt M^-1 * F)`
`u_g = u_n + _dt * v_(n-1/2) + _dt * avg_dt * M^-1 * F`
Then solve for F

`F = M*(u_g - u_n  - v_(n-1/2) * _dt)/(_dt * avg_dt)`
 `F = M*[(u_g-u_n) / (_dt*avg_dt) - v_(n-1/2)/avg_dt]`

Reply:
Thanks! The derivation makes sense to me. Then shouldn't `(computeQpValue() - _u[_qp])` be `(computeQpValue() - _u_old[_qp])` or am I missing something?

Reply:
And more importantly, shouldn't this be whole thing be multiplied by the mass matrix?

Reply:
The mass matrix is multiplied into the residual in the line after the highlighted section here. 

As for your first point, I think you are correct as u_n should equal the old solution at this point in the solve. Though using u_n = _u[_qp] still produces the right answer. There may an assignment of _u[_qp] that I am unaware of that is making the current formulation work by happenstance.

Using `_sys.solutionOld()(dofnum)` returns the same result

Reply:
I can understand why _u and _u_old hold the same value at the time of computing the BC residual. I recommend you to request _u_old (and also _u_dot_old) in the constructor. That way we don't rely on undefined (or undocumented) behavior.

Reply:
I know in `_fe_problem` you can set `u_dot_old`  but I don't believe there is a call to to requisition the old solution.

Reply:
Use `coupledDotOld`

Reply:
Are you saying to use `coupledDotOld(variable name)` instead of `_sys.solutionOld()(dofnum)` to call the old velocity?
That would involve an extra step as I would need to retrieve the correct variable name

Reply:
right... then just use `_var.uDotOld()`, `_var` is a reference to the underlying `MooseVariable`. Similarly use `_var.slnOld()` to get `_u_old`.

Reply:
Why is that better than this?
```
  Real resid = (computeQpValue() - _var.uDotOld()[dofnum]) / (avg_dt * _dt) -
               (_var.slnOld()[dofnum]) / avg_dt;
```

Reply:
Nevermind, that segfaults....

Reply:
Ah but the other method seg faults as well. The `dofnum` parameter is defined based on the total solution vector. I need a separate param for the local index

Reply:
Oh I am overthinking... I can just the _qp index.

Reply:
[8b672f9](https://github.com/idaholab/moose/pull/28175/commits/8b672f9a8c599fc0b8c3018bbd220daabc3c9884)